src/test/scala/edu/berkeley/cs/rise/opaque/TPCHTests.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package edu.berkeley.cs.rise.opaque
+
+
+import org.apache.spark.sql.SparkSession
+
+import edu.berkeley.cs.rise.opaque.benchmark._
+import edu.berkeley.cs.rise.opaque.benchmark.TPCH
+
+trait TPCHTests extends OpaqueTestsBase { self => 
+
+  def size = "sf_small"
+  def tpch: TPCH
+
+  override def beforeAll(): Unit = {
+    super.beforeAll();
+    tpch.ensureCached();
+  }
```

Comment:
This is great, thanks! Can you also call `generateMap` in here? Or put both `ensureCached` and `generateMap` in an init function.

Reply:
I can't because `generateMap` returns the actual map structure that the constructor needs. One alternative would be to do something like `val nameToDF = TPCHDataFrames.generateMap(sqlContext, Insecure, size, numPartitions)` outside of the auxiliary `this`constructor. The other is to set `nameToDF` to an arbitrary initialized value, then call an init function that sets it to the correct value while also calling `ensureCached`.

Reply:
The current code is just a bit confusing, since the default constructor is not meant to be used directly. I think `val nameToDF = TPCHDataFrames.generateMap(sqlContext, Insecure, size, numPartitions)` is good. Perhaps another way is to use a companion object and define `apply ` methods.

Reply:
I went with the `apply` method because that enables `generateMap` and `ensureCached` to be called in the same spot as a sort of init function. Thanks for the tip!